code/controllers/subsystem/spatial_gridmap.dm

```diff
@@ -0,0 +1,593 @@
+
+///if shit goes south, this will find hanging references for qdeleting movables inside the spatial grid
+/datum/controller/subsystem/spatial_grid/proc/find_hanging_cell_refs_for_movable(atom/movable/to_remove, remove_from_cells = TRUE)
+
+	var/list/queues_containing_movable = list()
+	for(var/queue_channel in waiting_to_add_by_type)
+		var/list/queue_list = waiting_to_add_by_type[queue_channel]
+		if(to_remove in queue_list)
+			queues_containing_movable += queue_channel//just add the associative key
+			if(remove_from_cells)
+				queue_list -= to_remove
+
+	if(!initialized)
+		return queues_containing_movable
+
+	var/list/containing_cells = list()
+	for(var/list/z_level_grid as anything in grids_by_z_level)
+		for(var/list/cell_row as anything in z_level_grid)
+			for(var/datum/spatial_grid_cell/cell as anything in cell_row)
+				if(to_remove in (cell.hearing_contents | cell.client_contents))
+					containing_cells += cell
+					if(remove_from_cells)
+						force_remove_from_cell(to_remove, cell)
+
+	return containing_cells
+
+///debug proc for checking if a movable is in multiple cells when it shouldnt be (ie always unless multitile entering is implemented)
+/atom/proc/find_all_cells_containing(remove_from_cells = FALSE)
+	var/datum/spatial_grid_cell/real_cell = SSspatial_grid.get_cell_of(src)
+	var/list/containing_cells = SSspatial_grid.find_hanging_cell_refs_for_movable(src, FALSE, remove_from_cells)
+
+	message_admins("[src] is located in the contents of [length(containing_cells)] spatial grid cells")
+
+	var/cell_coords = "the following cells contain [src]: "
+	for(var/datum/spatial_grid_cell/cell as anything in containing_cells)
+		cell_coords += "([cell.cell_x], [cell.cell_y], [cell.cell_z]), "
+
+	message_admins(cell_coords)
+	message_admins("[src] is supposed to only be contained in the cell at indexes ([real_cell.cell_x], [real_cell.cell_y], [real_cell.cell_z]). but is contained at the cells at [cell_coords]")
+
+///debug proc for finding how full the cells of src's z level are
+/atom/proc/find_grid_statistics_for_z_level(insert_clients = 0)
+	var/raw_clients = 0
+	var/raw_hearables = 0
+
+	var/cells_with_clients = 0
+	var/cells_with_hearables = 0
+
+	var/list/client_list = list()
+	var/list/hearable_list = list()
+
+	var/total_cells = (world.maxx / SPATIAL_GRID_CELLSIZE) ** 2
+
+	var/average_clients_per_cell = 0
+	var/average_hearables_per_cell = 0
+
+	var/hearable_min_x = (world.maxx / SPATIAL_GRID_CELLSIZE)
+	var/hearable_max_x = 1
+
+	var/hearable_min_y = (world.maxy / SPATIAL_GRID_CELLSIZE)
+	var/hearable_max_y = 1
+
+	var/client_min_x = (world.maxx / SPATIAL_GRID_CELLSIZE)
+	var/client_max_x = 1
+
+	var/client_min_y = (world.maxy / SPATIAL_GRID_CELLSIZE)
+	var/client_max_y = 1
+
+	var/list/inserted_clients = list()
+
+	if(insert_clients)
+		var/list/turfs
+		var/level = SSmapping.get_level(z)
+		if(is_station_level(level))
+			turfs = GLOB.station_turfs
+
+		else
+			turfs = block(locate(1,1,z), locate(255,255,z))
```

Comment:
oh one more thing, this shouldn't be hardcoded, right?

Reply:
probably not, fixed